[Serve/Doc] Update http with serve user guide (#27536)
ray-project/ray: doc/source/serve/http-guide.md

# HTTP with Serve

This section should help you:

> 

Reviewer: This section helps you to understand how to:
- configure HTTP server location
- send HTTP request to serve deployment
- use Ray Serve integration with FastAPI
- use customized HTTP Adapter

- Understand how to configure HTTP server location
- Understand how to send HTTP request to serve deployment
- Understand how Ray Serve integrate with FastAPI
- Understand how to use customized HTTP Adapter

## Configuring HTTP Server Locations

By default, Ray Serve starts a single HTTP server on the head node of the Ray cluster.
You can configure this behavior using the `http_options={"location": ...}` flag

> in {mod}`serve.start <ray.serve.start>`:

Reviewer 2: @sihanwang41 we shouldn't document `serve.start` here (in favor of `serve.run` now)

Reviewer 2: ping on this

Author: updated @edoakes

- "HeadOnly": start one HTTP server on the head node. Serve
  assumes the head node is the node you executed serve.start
  on. This is the default.
- "EveryNode": start one HTTP server per node.
- "NoServer" or `None`: disable HTTP server.

> 

Reviewer: Make these bold **HeadOnly**, **EveryNode**, **NoServer**, and remove the quotes.

:::{note}
Using the "EveryNode" option, you can point a cloud load balancer to the
instance group of Ray cluster to achieve high availability of Serve's HTTP

> proxies.

Reviewer: Use `EveryNode`. Also the main idea is HA, would it make sense if you rephrased this as:

To achieve high availability of Serve's HTTP proxies, use `EveryNode` option to point to the instance group of Ray cluster.

:::

(serve-http)=

## Calling Deployments via HTTP

When you create a deployment, it is exposed over HTTP by default at `/{deployment_name}`. You can change the route by specifying the `route_prefix` argument to the {mod}`@serve.deployment <ray.serve.api.deployment>` decorator.

> ```python

Reviewer: Put this in active voice with the subject.

When you create a deployment, Serve exposes your deployment over HTTP by default at ...

Any request to the Serve HTTP server at /counter is routed to the deployment’s __call__ method with a [Starlette Request object](https://www.starlette.io/requests/) as its sole argument.

# HTTP with Serve

This section should help you understand how to:

> - configure HTTP server location
> - send HTTP request to serve deployment
> - use Ray Serve integrate with FastAPI
> - use customized HTTP Adapter

Reviewer 3: ```suggestion
- configure your HTTP server location
- send HTTP requests to Serve deployments
- use Ray Serve to integrate with FastAPI
- use customized HTTP Adapters
```

## Configuring HTTP Server Locations

By default, Ray Serve starts a single HTTP server on the head node of the Ray cluster.
You can configure this behavior using the `http_options={"location": ...}` flag
in {mod}`serve.start <ray.serve.start>`:

> - **HeadOnly**: start one HTTP server on the head node. Serve
>   assumes the head node is the node you executed serve.run
>   on. This is the default.

Reviewer 3: ```suggestion
- **HeadOnly**: start one HTTP server on the head node. Serve
  assumes the head node is the node you executed `serve.run`
  on. This is the default.
```

- **EveryNode**: start one HTTP server per node.

> - **NoServer** or `None`: disable HTTP server.

Reviewer 2: sorry, I think you misunderstood my previous comment, this section should be totally dropped now because it's no longer configurable. We simply always go with the `EveryNode` behavior.

Reviewer 2: so let's get rid of that section and just specify that behavior as a top-level point

:::{note}

> To achieve high availability of Serve's HTTP proxies, use **EveryNode** option to point to the instance group of Ray cluster

Reviewer 3: ```suggestion
To achieve high availability of Serve's HTTP proxies, use the **EveryNode** option to point to the instance group of Ray cluster
```
What does instance group mean here? Maybe we should also mention using an external load balancer here, unless it'll be obvious to most readers

:::

(serve-http)=

## Calling Deployments via HTTP

> When you create a deployment, Serve exposes your deployment over HTTP by default at `/`. You can change the route by specifying the `route_prefix` argument to the {mod}`@serve.deployment <ray.serve.api.deployment>` decorator.

Reviewer 2: ```suggestion
When you deploy a Serve application, the ingress deployment (the one passed to `serve.run`) will be exposed over HTTP. If you want to route to another deployment, you can do so using the `ServeHandle` API <link to model comopsition>.
```

using `route_prefix` is no longer recommended! we want everything going through the single ingress

> When you create a deployment, Serve exposes your deployment over HTTP by default at `/{deployment_name}`. You can change the route by specifying the `route_prefix` argument to the {mod}`@serve.deployment <ray.serve.api.deployment>` decorator.

Reviewer 2: this is no longer true, now the "ingress deployment" is hosted on `/`

class FastAPIWrapper:
    pass

serve.run(FastAPIWrapper.bind())
```

This is useful for scaling out an existing FastAPI app with no modifications necessary.
Existing middlewares, automatic OpenAPI documentation generation, and other advanced FastAPI features should work as-is.

> To try it out, save a code snippet in a local python file (i.e. main.py) and in the same directory, run the following commands to start a local Ray cluster on your machine.

Reviewer 3: ```suggestion
To try it out, save a code snippet in a local python file (e.g. `main.py`) and in the same directory, run the following commands to start a local Ray cluster on your machine.
```

```python
async def json_resolver(request: starlette.requests.Request):
    return await request.json()
```

The input arguments to an HTTP adapter should be type-annotated. At a minimum, the adapter should accept a `starlette.requests.Request` type (https://www.starlette.io/requests/#request),
but it can also accept any type that's recognized by [FastAPI's dependency injection framework](https://fastapi.tiangolo.com/tutorial/dependencies/).

Here is an HTTP adapter that accepts two HTTP query parameters:

```python
def parse_query_args(field_a: int, field_b: str):
    return YourDataClass(field_a, field_b)
```

You can specify different type signatures to facilitate the extraction of HTTP fields, including
- [query parameters](https://fastapi.tiangolo.com/tutorial/query-params/),
- [body parameters](https://fastapi.tiangolo.com/tutorial/body/),
and 
- [many other data types](https://fastapi.tiangolo.com/tutorial/extra-data-types/).

For more details, you can take a look at the [FastAPI documentation](https://fastapi.tiangolo.com/).

You can use adapters in different scenarios within Serve, which we will go over one by one:

> 

Reviewer: May be make a transition here. 

In addition to above adapters, you also use other adapters. Below we examine at least three:

- Ray AIR `Predictor`
- Serve Deployment Graph `DAGDriver`
- Embedded in Bring Your Own `FastAPI` Application

### Ray AIR `Predictor`

Ray Serve provides a suite of adapters to convert HTTP requests to ML inputs like `numpy` arrays.
You can use them together with the [Ray AI Runtime (AIR) model wrapper](air-serving-guide) feature
to one-click deploy pre-trained models.

As an example, we provide a simple adapter for an *n*-dimensional array.

When using [model wrappers](air-serving-guide), you can specify your HTTP adapter via the `http_adapter` field:

```python
from ray import serve
from ray.serve.http_adapters import json_to_ndarray
from ray.serve import PredictorDeployment

serve.run(PredictorDeployment.options(name="my_model").bind(
    my_ray_air_predictor,
    my_ray_air_checkpoint,
    http_adapter=json_to_ndarray

> ))

Reviewer: Do you want to mention what `my_air_predictor`, `my_ray_air_checkpoint` are as arguments to the .`bind()` call?

```

### Serve Deployment Graph `DAGDriver`

When using a [Serve Deployment Graph](serve-deployment-graph), you can configure
`ray.serve.drivers.DAGDriver` to accept an HTTP adapter via its `http_adapter` field.

For example, the `json_request` adapter parses JSON in the HTTP body:

```python
from ray.serve.drivers import DAGDriver
from ray.serve.http_adapters import json_request
from ray.dag.input_node import InputNode

with InputNode() as input_node:
    # ...
    dag = DAGDriver.bind(other_node, http_adapter=json_request)
```

### Embedded in your existing `FastAPI` Application

You can also bring the adapter to your own FastAPI app using
[Depends](https://fastapi.tiangolo.com/tutorial/dependencies/#import-depends).
The input schema will automatically be part of the generated OpenAPI schema with FastAPI.

> 

Reviewer: The input schema automatically become part of the generated OpenAPI schema with FastAPI.

(just using present tense)

```python
from fastapi import FastAPI, Depends
from ray.serve.http_adapters import json_to_ndarray

app = FastAPI()

@app.post("/endpoint")
async def endpoint(np_array = Depends(json_to_ndarray)):
    ...
```

It has the following schema for input:

(serve-ndarray-schema)=

```{eval-rst}
.. autopydantic_model:: ray.serve.http_adapters.NdArray

```

### Pydantic models as adapters

Serve also supports [pydantic models](https://pydantic-docs.helpmanual.io/usage/models/) as a shorthand for HTTP adapters in model wrappers. Instead of using a function to define your HTTP adapter as in the examples above,

> you can directly pass in a pydantic model class to effectively tell Ray Serve "validate the HTTP body with this schema."

Reviewer: "to validate the ..."
